components/cosmetic_filters/resources/data/cosmetic_filters.js

```diff
@@ -0,0 +1,514 @@
+              notYetQueriedIds.push(mutatedId);
+              queriedIds.add(mutatedId);
+            };
+            break;
+        };
+      } else if (aMutation.addedNodes.length > 0) {
+        for (const node of aMutation.addedNodes) {
+          const element = asElement(node);
+          if (!element) {
+            continue;
+          };
+          const id = element.id;
+          if (id && !queriedIds.has(id)) {
+            notYetQueriedIds.push(id);
+            queriedIds.add(id);
+          };
+          const classList = element.classList;
+          if (classList) {
+            for (const className of classList.values()) {
+              if (className && !queriedClasses.has(className)) {
+                notYetQueriedClasses.push(className);
+                queriedClasses.add(className);
+              };
+            };
+          };
+        };
+      };
+    };
+    fetchNewClassIdRules();
+  };
+
+  const startObserving = () => {
+    // First queue up any classes and ids that exist before the mutation observer
+    // starts running.
+    const elmWithClassOrId = document.querySelectorAll('[class],[id]');
+    for (const elm of elmWithClassOrId) {
+      for (const aClassName of elm.classList.values()) {
+        queriedClasses.add(aClassName);
+      }
+      const elmId = elm.getAttribute('id');
+      if (elmId) {
+        queriedIds.add(elmId);
+      }
+    };
+
+    notYetQueriedClasses = Array.from(queriedClasses);
+    notYetQueriedIds = Array.from(queriedIds);
+    fetchNewClassIdRules();
+
+    // Second, set up a mutation observer to handle any new ids or classes
+    // that are added to the document.
+    cosmeticObserver = new MutationObserver(handleMutations);
+    let observerConfig = {
+      subtree: true,
+      childList: true,
+      attributeFilter: ['id', 'class']
+    };
+    if (document.documentElement instanceof Node) {
+      cosmeticObserver.observe(document.documentElement, observerConfig);
+    }
+  };
+
+  /**
+   * Provides a new function which can only be scheduled once at a time.
+   *
+   * @param onIdle function to run when the thread is less busy
+   * @param timeout max time to wait. at or after this time the function will be run regardless of thread noise
+  */
+  const idleize = (onIdle, timeout) => {
+    let idleId = undefined;
+    return function WillRunOnIdle () {
+      if (idleId !== undefined) {
+        return;
+      }
+      idleId = window.requestIdleCallback(() => {
+        idleId = undefined;
+        onIdle();
+      }, { timeout })
+    }
+  }
+
+  const _parseDomainCache = Object.create(null);
+  const getParsedDomain = (aDomain) => {
+    const cacheResult = _parseDomainCache[aDomain];
+    if (cacheResult !== undefined) {
+      return cacheResult;
+    }
+
+    const newResult = parseDomain(aDomain);
+    _parseDomainCache[aDomain] = newResult;
+    return newResult;
+  }
+
+  const isRelativeUrl = (url) => {
```

Comment:
should probably normalize url by converting it to lowercase and removing whitespace

also this would pass for non-http subresources like websockets (`wss://...`) and extension-injected resources (`chrome-extension://...`) and possibly some others.

Reply:
@SergeyZhukovsky for the latter, maybe just check that the url doesn't contain `://` ? we don't want third-party remote resources that aren't http/https to be treated as third parties.

Reply:
the problem with that as we accept URLs like `https:brave.com`. I'm going to use that to validate https://developer.mozilla.org/en-US/docs/Web/API/URL/URL